keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,699 @@
+-->
+
+#### How can an operator determine if the feature is in use by workloads?
+
+<!--
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+#### How can someone using this feature know that it is working for their instance?
+
+If a user terminates pods that are controlled by a deployment/job, then we should wait
+until the existing pods are terminated before starting new ones.
+
+When feature is turned on, we will also include a `terminating` field in the Job Status.
+
+We will add e2e test that determine this.
+
+#### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+<!--
+This is your opportunity to define what "normal" quality of service looks like
+for a feature.
+
+It's impossible to provide comprehensive guidance, but at the very
+high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99.9% of /health requests per day finish with 200 code
+
+These goals will help you determine what you need to measure (SLIs) in the next
+question.
+-->
+
+#### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+NA
+
+#### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+### Dependencies
+
+This feature is closely related to the 3329-retriable-and-nonretriable-failures but not sure if that is considered a dependency.
+
+#### Does this feature depend on any specific services running in the cluster?
+
+No
+
+### Scalability
+
+Generally, enabling this will slow down job creation if pods take a long time to terminate.  We would wait
+to create new pods until the existing ones are terminated
+
+#### Will enabling / using this feature result in any new API calls?
+
+No
+
+#### Will enabling / using this feature result in introducing new API types?
+
+We add `RecreatePodsWhen` to `JobSpec`.  This is a enum of two values.
+
+We add `terminating` to `JobStatus`. This is a pointer to type `int32`.
+
+#### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No
+
+#### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+For Job API, we are adding a enum field named `RecreatePodsWhen` which takes either a `TerminateOrFailed` or `Failed`
+
+- API type(s): enum
+- Estimated increase in size: 8B
+
+We are also added a status field for tracking terminating pods.
+
+- API type(s): int32
+- Estimated increase in size: 4B
+
+#### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+<!--
```

Comment:
This is important section, since like you mentioned before the additional wait time for pod termination will slow down the controller.

Reply:
updated.

Reply:
I don't think it makes sense to have SLOs for job completion, as this is outside of the control of the job controller. There can be SLOs for job sync (how fast it reacts to events). Although there aren't any official ones today.

Reply:
I see what you mean.